File: leetcode/0039.go
package leetcode

func combinationSum(candidates []int, target int) (ans [][]int) {
	comb := []int{}
	var dfs func(target, idx int)
	dfs = func(target, idx int) {
		if target == 0 {
			ans = append(ans, append([]int{}, comb...))
			return
		}
		if idx >= len(candidates) {
			return
		}
		dfs(target, idx+1)
		if target >= candidates[idx] {
			comb = append(comb, candidates[idx])
			dfs(target-candidates[idx], idx)
			comb = comb[:len(comb)-1]
		}
	}
	dfs(target, 0)
	return
}

func combinationSum1(candidates []int, target int) (ans [][]int) {
	comb := []int{}
	var dfs func(target, idx int)
	//idx 表示从此处选
	dfs = func(target, idx int) {
        if target == 0 {
            ans = append(ans,append([]int{},comb...))
            return
        }
        for i := idx; i < len(candidates)&&target >= candidates[idx];i++ {
            comb = append(comb,candidates[i])
            dfs(target-candidates[i],i)
            comb = comb[:len(comb)-1]
        }
	}
    sort.Ints(candidates)
	dfs(target, 0)
	return
}

File: leetcode/0039.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> tmp;
        vector<vector<int>> ans;
        int n = candidates.size();
        auto dfs = [&](auto&& dfs, auto target, auto i) {
            if (i == n) {
                return;
            }
            if (target == 0) {
                ans.emplace_back(tmp);
                return;
            }
            dfs(dfs, target, i + 1);
            if (target >= candidates[i]) {
                tmp.emplace_back(candidates[i]);
                dfs(dfs, target - candidates[i], i);
                tmp.pop_back();
            }
        };
        dfs(dfs, target, 0);
        return ans;
    }
};

class Solution1 {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> tmp;
        vector<vector<int>> ans;
        int n = candidates.size();
        auto dfs = [&](auto&& dfs, auto target, auto i) {
            if (target == 0) {
                ans.emplace_back(tmp);
                return;
            }
            for (auto j = i; j < n && target >= candidates[j]; j++) {
                tmp.emplace_back(candidates[j]);
                dfs(dfs, target - candidates[j], j);
                tmp.pop_back();
            }
        };
        sort(candidates.begin(), candidates.end());
        dfs(dfs, target, 0);
        return ans;
    }
};

class Solution {
public:
    vector<vector<int>> dp[1000], res;

    vector<vector<int>> combinationSum(vector<int>& c, int target) {
        dp[0].push_back(vector<int>());
        for (auto w : c)
            for (int j = w; j <= target; ++j)
                if (dp[j - w].empty() == 0) {
                    auto tmp = dp[j - w];
                    for (auto& ea : tmp) {
                        ea.push_back(w);
                        dp[j].push_back(ea);
                    }
                }
        res = dp[target];
        for (auto& i : res) {
            sort(i.begin(), i.end());
        }
        return res;
    }
};